File: sdk/python/ggid/saml.py
"""SAML Service Provider configuration and metadata generation.

Usage::

    from ggid.saml import SAMLConfig, generate_sp_metadata

    config = SAMLConfig(
        entity_id="https://myapp.example.com/saml",
        acs_url="https://myapp.example.com/saml/acs",
        slo_url="https://myapp.example.com/saml/slo",
    )
    metadata_xml = generate_sp_metadata(config)
"""

from __future__ import annotations
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from xml.sax.saxutils import escape


@dataclass
class SAMLConfig:
    """SAML Service Provider configuration."""
    entity_id: str
    acs_url: str
    slo_url: str = ""
    sign_requests: bool = False


def generate_sp_metadata(config: SAMLConfig) -> str:
    """Generate SAML SP metadata XML (EntityDescriptor).

    Args:
        config: SAMLConfig with entity_id and acs_url set.

    Returns:
        SP EntityDescriptor XML string.

    Raises:
        ValueError: If entity_id or acs_url is empty.
    """
    if not config.entity_id:
        raise ValueError("entity_id is required")
    if not config.acs_url:
        raise ValueError("acs_url is required")

    valid_until = (datetime.now(timezone.utc) + timedelta(days=365)).strftime(
        "%Y-%m-%dT%H:%M:%SZ"
    )

    parts = [
        '<?xml version="1.0" encoding="UTF-8"?>',
        f'<EntityDescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata"',
        f' entityID="{escape(config.entity_id)}"',
        f' validUntil="{valid_until}">',
        '<SPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">',
        "<NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress</NameIDFormat>",
        '<AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"',
        f' Location="{escape(config.acs_url)}" index="0"/>',
    ]

    if config.slo_url:
        parts.append(
            '<SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"'
            f' Location="{escape(config.slo_url)}"/>'
        )

    parts.append("</SPSSODescriptor>")
    parts.append("</EntityDescriptor>")
    return "".join(parts)


def fetch_idp_metadata(client) -> bytes:
    """Fetch IdP metadata XML from a GGID instance.

    Args:
        client: A ggid.Client instance with a configured base_url.

    Returns:
        IdP metadata XML bytes.
    """
    import urllib.request
    url = f"{client.base_url}/saml/idp/metadata"
    with urllib.request.urlopen(url) as resp:
        return resp.read()

File: sdk/ruby/lib/ggid/saml.rb
# GGID SDK - SAML SP utilities (Ruby)
module GGID
  module SAML
    module_function

    # Generate SAML SP metadata XML
    def generate_sp_metadata(entity_id:, acs_url:, slo_url: nil)
      slo = slo_url ?
        %(  <SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="#{escape_xml(slo_url)}" />\n) : ''
      %(<?xml version="1.0" encoding="UTF-8"?>
<EntityDescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" entityID="#{escape_xml(entity_id)}">
  <SPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
    <NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress</NameIDFormat>
    <AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
      Location="#{escape_xml(acs_url)}" index="0" />
#{slo}  </SPSSODescriptor>
</EntityDescriptor>)
    end

    # Fetch IdP metadata XML from GGID
    def fetch_idp_metadata(ggid_base_url)
      require 'net/http'
      require 'uri'
      uri = URI("#{ggid_base_url.gsub(/\/$/, '')}/saml/metadata")
      response = Net::HTTP.get_response(uri)
      raise "Failed to fetch IdP metadata: #{response.code}" unless response.is_a?(Net::HTTPSuccess)
      response.body
    end

    # Extract entity ID from IdP metadata XML
    def parse_entity_id(metadata_xml)
      match = metadata_xml.match(/entityID="([^"]+)"/)
      match && match[1]
    end

    # Extract SSO URL from IdP metadata XML
    def parse_sso_url(metadata_xml)
      match = metadata_xml.match(/SingleSignOnService[^>]*Location="([^"]+)"/)
      match && match[1]
    end

    # Build a SAML AuthnRequest redirect URL
    def build_authn_request_url(sso_url:, entity_id:, acs_url:, relay_state: nil)
      require 'base64'
      require 'cgi'
      id = "_#{Time.now.to_i}#{rand(36**8).to_s(36)}"
      request = %(<samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" ID="#{id}" Version="2.0" IssueInstant="#{Time.now.utc.iso8601}" Destination="#{escape_xml(sso_url)}" AssertionConsumerServiceURL="#{escape_xml(acs_url)}"><saml:Issuer xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">#{escape_xml(entity_id)}</saml:Issuer></samlp:AuthnRequest>)
      encoded = Base64.strict_encode64(request)
      sep = sso_url.include?('?') ? '&' : '?'
      url = "#{sso_url}#{sep}SAMLRequest=#{CGI.escape(encoded)}"
      url += "&RelayState=#{CGI.escape(relay_state)}" if relay_state
      url
    end

    def escape_xml(str)
      str.gsub('&', '&amp;').gsub('<', '&lt;').gsub('>', '&gt;').gsub('"', '&quot;').gsub("'", '&apos;')
    end
  end
end
